In [1]:
import pandas as pd

In [2]:
!git clone https://github.com/taehojo/data.git

Cloning into 'data'...
remote: Enumerating objects: 21, done.
remote: Counting objects: 100% (21/21), done.
remote: Compressing objects: 100% (18/18), done.
remote: Total 21 (delta 3), reused 20 (delta 2), pack-reused 0
Unpacking objects: 100% (21/21), done.


In [15]:
df = pd.read_csv('./data/sonar3.csv', header = None)
df.head()

# 0번부터 59번까지 음파 주파수의 에너지를 0에서 1사이의 숫자로 표현
# 60번 : 일반 암석 0, 광석 1

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   0  
1  0.0052  0.0044   0  
2  0.0095  0.0078   0  
3  0.0040  0.0117   0  
4  0.0107  0.0094   0  

[5 rows x 61 columns]

In [16]:
# y 값 확인
df[60].value_counts()

# 광석 111 / 암석 97

1    111
0     97
Name: 60, dtype: int64

In [17]:
# 정보와 결측치 확인
df.info()
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       208 non-null    float64
 1   1       208 non-null    float64
 2   2       208 non-null    float64
 3   3       208 non-null    float64
 4   4       208 non-null    float64
 5   5       208 non-null    float64
 6   6       208 non-null    float64
 7   7       208 non-null    float64
 8   8       208 non-null    float64
 9   9       208 non-null    float64
 10  10      208 non-null    float64
 11  11      208 non-null    float64
 12  12      208 non-null    float64
 13  13      208 non-null    float64
 14  14      208 non-null    float64
 15  15      208 non-null    float64
 16  16      208 non-null    float64
 17  17      208 non-null    float64
 18  18      208 non-null    float64
 19  19      208 non-null    float64
 20  20      208 non-null    float64
 21  21      208 non-null    float64
 22  22

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [18]:
# 결과 데이터 분리하기
X = df.drop([60], axis=1)
y = df[60]

display(X)
display(y)

0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53      54      55      56  \
0    0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1    0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2    0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3    0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4    0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   
..      ...  ...     ...     ...     ...     ...     ...     ...     ...   
203  0.2684  ...  0.0203  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065   
204  0.2154  ...  0.0051  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034   
205  0.2529  ...  0.0155  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140   
206  0.2354  ...  0.0042  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034   
207  0.2354  ...  0.0181  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040   

         57      58      59  
0    0.0084  0.0090  0.0032  
1    0.0049  0.0052  0.0044  
2    0.0164  0.0095  0.0078  
3    0.0044  0.0040  0.0117  
4    0.0048  0.0107  0.0094  
..      ...     ...     ...  
203  0.0115  0.0193  0.0157  
204  0.0032  0.0062  0.0067  
205  0.0138  0.0077  0.0031  
206  0.0079  0.0036  0.0048  
207  0.0036  0.0061  0.0115  

[208 rows x 60 columns]

0      0
1      0
2      0
3      0
4      0
      ..
203    1
204    1
205    1
206    1
207    1
Name: 60, Length: 208, dtype: int64

In [19]:
# 신경망 모델을 관리하는 객체
from tensorflow.keras.models import Sequential
# 선형 회귀 레이어
from tensorflow.keras.layers import Dense
# 활성화 함수를 정의하는 객체
from tensorflow.keras.layers import Activation

# 모델 설정
# 신경망 설계 
model = Sequential()
# 은닉층
model.add(Dense (24, input_dim = 60))
# 활성화 함수
model.add(Activation('relu'))

model.add(Dense(10, activation = 'relu'))

# 출력층
model.add(Dense(1, activation = 'sigmoid'))

# 모델 컴파일
# 2진분류! : binary_crossentropy
model.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics=['accuracy'])
model.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_12 (Dense)            (None, 24)                1464      
                                                                 
 activation_4 (Activation)   (None, 24)                0         
                                                                 
 dense_13 (Dense)            (None, 10)                250       
                                                                 
 dense_14 (Dense)            (None, 1)                 11        
                                                                 
Total params: 1,725
Trainable params: 1,725
Non-trainable params: 0
_________________________________________________________________


In [20]:
# 모델 실행
model.fit(X, y, epochs=200, batch_size=10)

Epoch 1/200
21/21 [==============================] - 1s 2ms/step - loss: 0.7001 - accuracy: 0.4808
Epoch 2/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6808 - accuracy: 0.5192
Epoch 3/200
21/21 [==============================] - 0s 3ms/step - loss: 0.6564 - accuracy: 0.6010
Epoch 4/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6358 - accuracy: 0.6971
Epoch 5/200
21/21 [==============================] - 0s 3ms/step - loss: 0.6187 - accuracy: 0.7019
Epoch 6/200
21/21 [==============================] - 0s 2ms/step - loss: 0.6045 - accuracy: 0.6923
Epoch 7/200
21/21 [==============================] - 0s 2ms/step - loss: 0.5878 - accuracy: 0.7548
Epoch 8/200
21/21 [==============================] - 0s 2ms/step - loss: 0.5721 - accuracy: 0.7356
Epoch 9/200
21/21 [==============================] - 0s 3ms/step - loss: 0.5551 - accuracy: 0.7596
Epoch 10/200
21/21 [==============================] - 0s 2ms/step - loss: 0.5391 - accuracy: 0.7644
Epoch 11/

In [21]:
# 정확도 확인
a = model.evaluate(X, y)
a
print(f'손실률 : {a[0]}')
print(f'정확도 : {a[1]}')

7/7 [==============================] - 0s 3ms/step - loss: 0.0140 - accuracy: 1.0000
손실률 : 0.014023312367498875
정확도 : 1.0




---



```
결과 정확도가 100%

이게 광물 데이터 인데 
진짜 100%의 확률일까...ㅋㅋㅋㅋ
나도 나를 못믿는데 이건 더 아리쏭하네 
```

그래서 나온게 과적합 !

- 과적합 모델이란? 모델이 학습데이터안에서 일정 수준 이상의 예측 정확도를 보여주지만,<br> 새로운 데이터에 적용하면 잘 맞지 않는 것을 의미.
- 층이 너무 많거나 변수가 복잡해서 발생하기도 함.
- 테스트 셋과 학습셋이 중복될때 생기기도 함.

- 딥러닝은 학습단계에서 입력층, 은닉층, 출력층의 노드를에 상당히 많은 변수가 투입되기 때문에 과적합에 빠지지 않게 주의 해야함.




---



✔ 학습 데이터와 테스트 데이터 분리 

In [36]:
# X, y데이 각각 정해진 비율(%)만큼 학습셋과 테스트 셋으로 분리시키는 사이킷런의 함수
# 7:3으로 
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.3, shuffle = True)

모델은 앞에 구성과 동일하게 하지만 위로 올려야하니 다시 적는걸로 

In [37]:
X_train.shape, y_train.shape, X_test.shape

((145, 60), (145,), (63, 60))

In [38]:
# 신경망 모델을 관리하는 객체
from tensorflow.keras.models import Sequential
# 선형 회귀 레이어
from tensorflow.keras.layers import Dense
# 활성화 함수를 정의하는 객체
from tensorflow.keras.layers import Activation

# 모델 설정
# 신경망 설계 
model = Sequential()
# 은닉층
model.add(Dense (24, input_dim = 60))
# 활성화 함수
model.add(Activation('relu'))

model.add(Dense(10, activation = 'relu'))

# 출력층
model.add(Dense(1, activation = 'sigmoid'))

# 모델 컴파일
# 2진분류! : binary_crossentropy
model.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics=['accuracy'])
model.summary()

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_21 (Dense)            (None, 24)                1464      
                                                                 
 activation_7 (Activation)   (None, 24)                0         
                                                                 
 dense_22 (Dense)            (None, 10)                250       
                                                                 
 dense_23 (Dense)            (None, 1)                 11        
                                                                 
Total params: 1,725
Trainable params: 1,725
Non-trainable params: 0
_________________________________________________________________


In [39]:
# 모델 실행
model.fit(X_train, y_train, epochs=200, batch_size=10)

Epoch 1/200
15/15 [==============================] - 1s 3ms/step - loss: 0.6978 - accuracy: 0.5655
Epoch 2/200
15/15 [==============================] - 0s 3ms/step - loss: 0.6693 - accuracy: 0.6966
Epoch 3/200
15/15 [==============================] - 0s 3ms/step - loss: 0.6587 - accuracy: 0.6966
Epoch 4/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6507 - accuracy: 0.7448
Epoch 5/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6408 - accuracy: 0.7655
Epoch 6/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6353 - accuracy: 0.7448
Epoch 7/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6264 - accuracy: 0.7793
Epoch 8/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6202 - accuracy: 0.7379
Epoch 9/200
15/15 [==============================] - 0s 2ms/step - loss: 0.6034 - accuracy: 0.7448
Epoch 10/200
15/15 [==============================] - 0s 2ms/step - loss: 0.5912 - accuracy: 0.8000
Epoch 11/

In [40]:
# 정확도 확인
b = model.evaluate(X_test, y_test)
print(f'손실률 : {b[0]}')
print(f'정확도 : {b[1]}')

2/2 [==============================] - 0s 7ms/step - loss: 1.0063 - accuracy: 0.7460
손실률 : 1.0063157081604004
정확도 : 0.7460317611694336


※ 학습데이터   X_train, y_train 에서 나온 정확도 95.86% <br>
※ 테스트데이터 X_test, y_test   에서 나온 정확도 74.60%

✔ 모델 성능 향상을 위한 방법
1. 데이터를 보강하는 방법
2. 알고리즘을 최적화 하는 방법

In [42]:
# 학습 후 만든 모델 저장
model.save('./my_model.hdf5')
# hdf5 파일포맷은 주로 과학 기술데이터 작업에서 사용
# 크고 복잡한 데이터를 저장하는데 사용
# 모델 불러올때는 케라스 api load_model 함수 사용.

from tensorflow.keras.models import Sequential, load_model

# 불러온 모델 메모리 삭제
del model

In [50]:
# 저장한 모델 불러오기
model = load_model('/content/my_model.hdf5')

# 불러온 모델을 적용해 정확도 확인.
score = model.evaluate(X_test, y_test)
print(f'손실률 : {score[0]}')
print(f'정확도 : {score[1]}')

2/2 [==============================] - 0s 6ms/step - loss: 1.0063 - accuracy: 0.7460
손실률 : 1.0063157081604004
정확도 : 0.7460317611694336




---



전체 데이터에서 7:3으로 설정할 경우 30%의 테스트셋은 학습에 사용할 수 없다는 단점.<br>

이를 해결하고자 
```k겹 교차검증(K-fold cross calidation)```<br> 
K겹 교차 검증이란?<br>
    - 데이터셋을 여러 개로 나누어 하나씩 테스트셋으로 사용하고 나머지를 모두 합해서 학습셋으로 사용하는 방법.<br>
    이렇게 하면 데이터 100%를 학습데이터로 사용할 수 있고, 테스트데이터로도 사용할 수 있음.

In [59]:
# 사이킷런 KFold() 함수 
from sklearn.model_selection import KFold

from sklearn.metrics import accuracy_score

# 몇개의 파일로 나눌것인가 ? 5
k = 5

#KFold 함수를 불러옵니다. 분할하기 전에 샘플이 치우치지 않도록 섞어 줍니다.
kfold = KFold(n_splits=k, shuffle=True)

# 정확도 채울 리스트
acc_score = []

def model_fn():
    model = Sequential()
    model.add(Dense(24, input_dim = 60, activation = 'relu'))
    model.add(Dense(10, activation = 'relu'))
    model.add(Dense(1, activation = 'sigmoid'))

    return model

# k개의 학습데이터, 테스트데이터로 분리되면서 반복
for train_index , test_index in kfold.split(X): 
    X_train , X_test = X.iloc[train_index,:], X.iloc[test_index,:]  
    y_train , y_test = y.iloc[train_index], y.iloc[test_index]

    model = model_fn()
    model.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics = ['accuracy'])
    history = model.fit(X_train, y_train, epochs = 200, batch_size=10, verbose=0)

    # 정확도 
    accuracy = model.evaluate(X_test, y_test)[1]
    # 리스트에 저장
    acc_score.append(accuracy)

avg_score = sum(acc_score) / k

# 제발ㅋㅋㅋ 철자좀ㅋㅋㅋㅋ 틀리지 말자

2/2 [==============================] - 0s 10ms/step - loss: 0.7050 - accuracy: 0.8537


In [60]:
print(f'정확도 : {acc_score}')
print(f'정확도 평균 : {avg_score}')

# 5번의 정확도

정확도 : [0.8571428656578064, 0.8095238208770752, 0.761904776096344, 0.7804877758026123, 0.8536585569381714]
정확도 평균 : 0.8125435590744019
